### ISSUE TITLE
Performance bug with convolutions with weights and inputs of similar spatial size

### ISSUE THREAD

**Issue by LaurenceA** [pytorch/pytorch#35603]
This is a performance bug in `conv2d`, when spatial dimensions of the weight and input are the same (or similar).  Such convolutions get 5 to 10 times slower when transitioning from image sizes of 30 to 32.  There may be an issue with the selection of the CUDNN convolution algorithm.

One possibly easy route to a solution is to allow manual selection of the convolution algorithm.

## To Reproduce

```
import torch as t
t.backends.cudnn.benchmark=True
import torch.nn.functional as F
from timeit import default_timer as timer

device = "cuda"
Cin = 128
Cout = 256
B = 1000

for W in range(20, 40):
    input = t.randn(B, Cin, W, W, device=device)
    weight = t.randn(Cout, Cin, W, W, device=device)
    output = F.conv2d(input, weight, padding=1)
    t.cuda.synchronize(device=device)

    start_time = timer()
    output = F.conv2d(input, weight, padding=1)
    t.cuda.synchronize(device=device)
    print(f"W: {W}, W**2: {W**2:4d}, time: {1000*(timer()-start_time)}")
```

## Expected behavior

The time required should scale linearly with `W**2`, instead, there is around a factor of 5 jump between W=30 and W=32,
```
W: 20, W**2:  400, time: 14.050104655325413
W: 21, W**2:  441, time: 13.91313411295414
W: 22, W**2:  484, time: 13.899956829845905
W: 23, W**2:  529, time: 13.991509564220905
W: 24, W**2:  576, time: 14.103865250945091
W: 25, W**2:  625, time: 14.099820517003536
W: 26, W**2:  676, time: 14.254673384130001
W: 27, W**2:  729, time: 14.209367334842682
W: 28, W**2:  784, time: 14.332166872918606
W: 29, W**2:  841, time: 14.398249797523022
W: 30, W**2:  900, time: 14.406020753085613
W: 31, W**2:  961, time: 53.21667902171612
W: 32, W**2: 1024, time: 86.8472745642066
W: 33, W**2: 1089, time: 84.1836640611291
W: 34, W**2: 1156, time: 113.77124954015017
W: 35, W**2: 1225, time: 84.53960344195366
W: 36, W**2: 1296, time: 125.90546812862158
W: 37, W**2: 1369, time: 84.77605786174536
W: 38, W**2: 1444, time: 141.37933682650328
W: 39, W**2: 1521, time: 149.50636308640242
```
The performance with `benchmark=True` commented out is even more exciting,
```
W: 20, W**2:  400, time: 17.404177226126194
W: 21, W**2:  441, time: 17.446410842239857
W: 22, W**2:  484, time: 17.630738206207752
W: 23, W**2:  529, time: 17.792406491935253
W: 24, W**2:  576, time: 16.371975652873516
W: 25, W**2:  625, time: 13.557913713157177
W: 26, W**2:  676, time: 13.596143573522568
W: 27, W**2:  729, time: 13.673634268343449
W: 28, W**2:  784, time: 13.687632977962494
W: 29, W**2:  841, time: 13.748884201049805
W: 30, W**2:  900, time: 13.820515014231205
W: 31, W**2:  961, time: 53.05525008589029
W: 32, W**2: 1024, time: 118.49690321832895
W: 33, W**2: 1089, time: 84.05695110559464
W: 34, W**2: 1156, time: 84.2417012900114
W: 35, W**2: 1225, time: 84.5922939479351
W: 36, W**2: 1296, time: 84.85969807952642
W: 37, W**2: 1369, time: 85.06182674318552
W: 38, W**2: 1444, time: 84.93016473948956
W: 39, W**2: 1521, time: 85.2658236399293
```

This is pretty painful, because it means my algorithm is very fast on MNIST, but disproportionately slower on any other dataset, such as CIFAR10.

## Environment

```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: NVS 510
GPU 1: TITAN V
GPU 2: TITAN V

Nvidia driver version: 418.74
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] numpydoc==0.9.1
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.14           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] pytorch                   1.3.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.4.1                py37_cu101    pytorch
```

**Comment by ptrblck** [comment]
I can reproduce similar timings using cudnn7.6.5.32.

cudnn kernels are mostly optimized for "standard" use cases adapted from commonly used models, which might not be the case for this setup.
Selecting the algorithm manually would thus most likely not yield any performance gains, as `cudnn.benchmark=True` already yields the fastest performance compared to the default algo and non-cudnn kernels.

I'll forward the issue to the cudnn team.

**Comment by LaurenceA** [comment]
Thanks.  It's worth pointing out that there must be an algorithm selection issue somewhere, because it reliably runs more slowly with `cudnn.benchmark=True` for certain image sizes (here, `W=34, 36, 38, 39`)

**Comment by ptrblck** [comment]
With warmup iterations and an average over 100 iterations for profiling, I get better (or nearly identical) results with `benchmark=True` on cudnn7.6.5.32:

```python
# benchmark = False
W: 20, W**2:  400, time: 14.08484622836113
W: 21, W**2:  441, time: 14.180783098563552
W: 22, W**2:  484, time: 14.277910813689232
W: 23, W**2:  529, time: 14.383608615025878
W: 24, W**2:  576, time: 14.480862868949771
W: 25, W**2:  625, time: 12.220423808321357
W: 26, W**2:  676, time: 12.271838188171387
W: 27, W**2:  729, time: 12.29959312826395
W: 28, W**2:  784, time: 12.350493092089891
W: 29, W**2:  841, time: 12.40236908197403
W: 30, W**2:  900, time: 12.449518097564578
W: 31, W**2:  961, time: 48.68807701393962
W: 32, W**2: 1024, time: 109.0675581805408
W: 33, W**2: 1089, time: 73.0636949185282
W: 34, W**2: 1156, time: 73.28957760706544
W: 35, W**2: 1225, time: 73.35267836228013
W: 36, W**2: 1296, time: 73.5642549674958
W: 37, W**2: 1369, time: 73.7246550898999
W: 38, W**2: 1444, time: 73.88410283252597
W: 39, W**2: 1521, time: 74.08880545757711

# benchmark = True
W: 20, W**2:  400, time: 12.098588841035962
W: 21, W**2:  441, time: 12.14685458689928
W: 22, W**2:  484, time: 12.146899402141571
W: 23, W**2:  529, time: 12.17234268784523
W: 24, W**2:  576, time: 12.21048871986568
W: 25, W**2:  625, time: 12.25636157207191
W: 26, W**2:  676, time: 12.3215610999614
W: 27, W**2:  729, time: 12.348472103476524
W: 28, W**2:  784, time: 12.412199433892965
W: 29, W**2:  841, time: 12.46114069595933
W: 30, W**2:  900, time: 12.496461411938071
W: 31, W**2:  961, time: 48.89828487299383
W: 32, W**2: 1024, time: 66.92298295907676
W: 33, W**2: 1089, time: 73.23627979494631
W: 34, W**2: 1156, time: 73.29694145359099
W: 35, W**2: 1225, time: 73.5847204271704
W: 36, W**2: 1296, time: 73.65927060134709
W: 37, W**2: 1369, time: 73.7915328796953
W: 38, W**2: 1444, time: 73.98397096432745
W: 39, W**2: 1521, time: 74.16641254909337
```
Code:
```python
import torch as t
t.backends.cudnn.benchmark=True
import torch.nn.functional as F
from timeit import default_timer as timer

device = "cuda"
Cin = 128
Cout = 256
B = 1000
nb_iters = 100


for W in range(20, 40):
    input = t.randn(B, Cin, W, W, device=device)
    weight = t.randn(Cout, Cin, W, W, device=device)
    for _ in range(20):
        output = F.conv2d(input, weight, padding=1)

    t.cuda.synchronize(device=device)
    start_time = timer()
    for _ in range(nb_iters):
        output = F.conv2d(input, weight, padding=1)
    t.cuda.synchronize(device=device)
    print(f"W: {W}, W**2: {W**2:4d}, time: {1000*(timer()-start_time)/nb_iters}")
```

**Comment by LaurenceA** [comment]
Interesting: thanks!

Importantly, this still replicates the main issue: the factor of 5-6
decrease in time taken as we move from W=30 to W=32.

**Comment by LaurenceA** [comment]
I'm coming back to this.  Turns out you can use these similar-sized input and weight convolutions to compute covariance matrices for the inputs in a convolutional layer.  This has applications in at least a couple of interesting application areas:
* Natural gradient (KFAC) optimisation.
* Bayesian inference with Laplace (e.g. https://github.com/aleximmer/Laplace)
